clippy_lints/src/replace_box.rs

```diff
@@ -33,16 +36,52 @@ declare_clippy_lint! {
     perf,
     "assigning a newly created box to `Box<T>` is inefficient"
 }
-declare_lint_pass!(ReplaceBox => [REPLACE_BOX]);
+
+#[derive(Default)]
+pub struct ReplaceBox {
+    // Stack of caches for moved vars. The latest entry is the
+    // body being currently visited.
+    moved_vars_caches: Vec<Option<HirIdSet>>,
+}
+
+impl ReplaceBox {
+    fn current_moved_vars_cache(&mut self) -> &mut Option<HirIdSet> {
+        self.moved_vars_caches.last_mut().unwrap()
+    }
+}
+
+impl_lint_pass!(ReplaceBox => [REPLACE_BOX]);
 
 impl LateLintPass<'_> for ReplaceBox {
+    fn check_body(&mut self, _: &LateContext<'_>, _: &Body<'_>) {
+        self.moved_vars_caches.push(None);
+    }
+
+    fn check_body_post(&mut self, _: &LateContext<'_>, _: &Body<'_>) {
+        _ = self.moved_vars_caches.pop();
+    }
+
     fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ Expr<'_>) {
         if let ExprKind::Assign(lhs, rhs, _) = &expr.kind
             && !lhs.span.from_expansion()
             && !rhs.span.from_expansion()
             && let lhs_ty = cx.typeck_results().expr_ty(lhs)
             // No diagnostic for late-initialized locals
-            && lhs.res_local_id().is_none_or(|local| local_is_initialized(cx, local))
+            && let local = lhs.res_local_id()
+            && local.is_none_or(|local| {
+                local_is_initialized(cx, local)
+                    && !self.current_moved_vars_cache().get_or_insert({
```

Comment:
This should probably be using `get_or_insert_with`. Currently all of this is eagerly evaluated, making the cache essentially unused

Reply:
Yeah, good catch indeed!